ipq40xx: allow use of ancillary partitions in WHW03 V2

Lanchon opened the pull request:
As provisioned by the OEM, WHW03 V2 has dual firmware sets plus a shared partition for data storage (mtd14: "syscfg", 181 MiB).

This partition is the largest one of the device, and in OpenWrt it would be useful as persistent storage surviving sysupgrade, but unfortunately it is declared read-only.

There is also a smaller unused partition in the same situation (mtd13: "sysdiag", 2 MiB),

### Version 1:

This commit makes the two aforementioned partitions writable, for the benefit of users who may want to use them.

### Version 2:

This commit repurposes the space of the two aforementioned unused partitions to define a new "extra" partition for users who may want to take advantage of the extra storage.

Also, the stock partitions are kept accessible and writable, to assist in returning to stock firmware if ever desired.

Lanchon commented:
@vtremblay @hauke @Ansuel

Ansuel commented:
@Lanchon are these 2 partition recreated/reinit on reverting to the factory image?

We can consider just using this extra space and split it to the 2 partition... (if the bootloader is ok with this)

Lanchon committed:
ipq40xx: allow use of ancillary partitions in WHW03 V2

As provisioned by the OEM, WHW03 V2 has dual firmware sets plus
a shared partition for data storage (mtd14: "syscfg", 181 MiB).

This partition is the largest one of the device, and in OpenWrt
it would be useful as persistent storage surviving sysupgrade,
but unfortunately it is declared read-only.

There is also a smaller unused partition in the same situation
(mtd13: "sysdiag", 2 MiB),

This commit repurposes the space of the two aforementioned
unused partitions to define a new "extra" partition for users
who may want to take advantage of the extra storage.

Also, the stock partitions are kept accessible and writable,
to assist in returning to stock firmware if ever desired.

Signed-off-by: Rodrigo Balerdi <lanchon@gmail.com>

Lanchon commented:
hi @Ansuel, thanks!

i fixed the sign-off thing and moved to draft since there is some space for discussion.

> are these 2 partition recreated/reinit on reverting to the factory image?

the small one is just erased (0xFFs). the larger one gets wiped when "factory resetting", i would guess, due to its function.

however it's a bit of a moot point: openwrt's sysupgrade does the right thing and flashes on the inactive boot partition set, then toggles `boot_part` (`1`/`2`). so the moment you sysupgrade in openwrt, stock firmware is gone. and i've been trying to return to stock, to test this thing you ask among other things, and wasn't able to (see https://github.com/openwrt/openwrt/commit/afcf1a4de4599a371f14db170b8456da915088bf#commitcomment-130567754). so what's the use of worrying about stock? on top, it's no like openwrt is using these partitions, i just want to let advanced users use them if they want.

> We can consider just using this extra space and split it to the 2 partition... (if the bootloader is ok with this)

as i see it, we have 4 options, 2 simple and 2 complex:

```
[    1.277588] Creating 15 MTD partitions on "qcom_nand.0":
[    1.284466] 0x000000000000-0x000000100000 : "SBL1"
[    1.291446] 0x000000100000-0x000000200000 : "MIBIB"
[    1.296083] 0x000000200000-0x000000300000 : "QSEE"
[    1.300798] 0x000000300000-0x000000380000 : "CDT"
[    1.305094] 0x000000380000-0x000000580000 : "APPSBL"
[    1.311471] 0x000000580000-0x000000600000 : "ART"
[    1.314994] 0x000000600000-0x000000680000 : "u_env"
[    1.319582] 0x000000680000-0x0000006c0000 : "s_env"
[    1.324016] 0x0000006c0000-0x000000700000 : "devinfo"
[    1.328823] 0x000000700000-0x00000a800000 : "kernel"
[    1.490871] 0x000000d00000-0x00000a800000 : "rootfs"
[    1.643151] mtd: setting mtd10 (rootfs) as root device
[    1.643537] mtdsplit: no squashfs found in "rootfs"
[    1.647219] 0x00000a800000-0x000014900000 : "alt_kernel"
[    1.809863] 0x00000ae00000-0x000014900000 : "alt_rootfs"
[    1.959552] 0x000014900000-0x000014b00000 : "sysdiag"
[    1.961963] 0x000014b00000-0x000020000000 : "syscfg"
```

1) make "sysdiag" and "syscfg" writable. we get one 181 MiB partition and a useless 2 MiB part. i put this option in draft mode for now

2) my recommended:
```
0x000014900000-0x000014b00000 : "stock_sysdiag"
0x000014b00000-0x000020000000 : "stock_syscfg"
0x000014900000-0x000020000000 : "extra"
```
we get one 183 MiB part, and access to return to stock if desired. i used this `stock_*` paradigm in https://openwrt.org/inbox/toh/arris/tr4400_v2, which required extensive partition rework to be able to coalesce most free space into a big UBI, including moving the irreplaceable `fw_env` paritition containing MACs, etc:
https://github.com/openwrt/openwrt/blob/bb8fd41f9a6775734882fe9707f7ad5e5edfbf9e/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts#L255-L258
https://github.com/openwrt/openwrt/blob/bb8fd41f9a6775734882fe9707f7ad5e5edfbf9e/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts#L272-L295

some space was not able to be coalesced because it had bootloader stuff in the way, so i defined an `extra` partition for it:
https://github.com/openwrt/openwrt/blob/bb8fd41f9a6775734882fe9707f7ad5e5edfbf9e/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts#L202-L205
https://github.com/openwrt/openwrt/blob/bb8fd41f9a6775734882fe9707f7ad5e5edfbf9e/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts#L300-L303

in all cases, although the part names were prefixed by `stock_`, the MTD numbers were conserved, so partitions could be accessed by scripts by number in both stock and openwrt firmwares.

again, i recommend this which sticks to these guidelines:
```
0x000014900000-0x000014b00000 : "stock_sysdiag"
0x000014b00000-0x000020000000 : "stock_syscfg"
0x000014900000-0x000020000000 : "extra"
```

3) adding the extra space to boot part set 2:

> We can consider just using this extra space and split it to the 2 partition... (if the bootloader is ok with this)

the firmware should be ok, as the kernel is raw outside of UBI, but the FW env and the openwrt build might need to be modified.

FW env:
```
altkern=a800000
auto_recovery=yes
baudrate=115200
boot_part_ready=3
bootcmd=if test $auto_recovery = no; then bootipq; elif test $boot_part = 1; then run bootpart1; else run bootpart2; fi
bootdelay=2
bootpart1=set bootargs $partbootargs && nand read $loadaddr $prikern $kernsize && bootm $loadaddr
bootpart2=set bootargs $partbootargs2 && nand read $loadaddr $altkern $kernsize && bootm $loadaddr
fdt_high=87000000
flash_type=2
flashimg=tftpboot $loadaddr $image && nand erase $prikern $imgsize && nand write $loadaddr $prikern $filesize
flashimg2=tftpboot $loadaddr $image && nand erase $altkern $imgsize && nand write $loadaddr $altkern $filesize
image=nodes_v2.img
imgsize=a100000
ipaddr=192.168.1.11
kernsize=d00000
loadaddr=81000000
machid=8010006
maxpartialboots=3
partbootargs=init=/sbin/init rootfstype=ubifs ubi.mtd=rootfs root=ubi0:ubifs rootwait rw
partbootargs2=init=/sbin/init rootfstype=ubifs ubi.mtd=alt_rootfs root=ubi0:ubifs rootwait rw 
prikern=700000
serverip=192.168.1.236
stderr=serial
stdin=serial
stdout=serial
boot_part=2
```

the problem here is that, during sysupgrade, kernel and UBI are written to the unused firmware slot and then `boot_part` is toggled. so what is the use of enlarging one UBI part while its alter ego stays the previous size? and disabling auto recovery means altering the fw variables, so this is a "complex" change.

4) if a complex change is desired (changing the build AND the FW env), then i would argue this is better:

change:

```
0x000000700000-0x00000a800000 : "kernel"
0x000000d00000-0x00000a800000 : "rootfs"
0x00000a800000-0x000014900000 : "alt_kernel"
0x00000ae00000-0x000014900000 : "alt_rootfs"
0x000014900000-0x000014b00000 : "sysdiag"
0x000014b00000-0x000020000000 : "syscfg"
```
to:
```
0x000000700000-0x00000a800000 : "stock_kernel"
0x000000d00000-0x00000a800000 : "stock_rootfs"
0x00000a800000-0x000014900000 : "stock_alt_kernel"
0x00000ae00000-0x000014900000 : "stock_alt_rootfs"
0x000014900000-0x000014b00000 : "stock_sysdiag"
0x000014b00000-0x000020000000 : "stock_syscfg"
0x000000700000-0x000001300000 : "recovery" - 12 MiB
0x000001300000-0x000001F00000 : "kernel" - 12 MiB
0x000001F00000-0x000020000000 : "ubi" - 481 MiB
```

Lanchon commented:
hi @Ansuel, thanks!

i fixed the sign-off thing and moved to draft since there is some space for discussion.

> are these 2 partition recreated/reinit on reverting to the factory image?

the small one is just erased (0xFFs). the larger one gets wiped when "factory resetting", i would guess, due to its function.

however it's a bit of a moot point: openwrt's sysupgrade does the right thing and flashes on the inactive boot partition set, then toggles `boot_part` (`1`/`2`). so the moment you sysupgrade in openwrt, stock firmware is gone. and i've been trying to return to stock, to test this thing you ask among other things, and wasn't able to (see https://github.com/openwrt/openwrt/commit/afcf1a4de4599a371f14db170b8456da915088bf#commitcomment-130567754). so what's the use of worrying about stock? on top, it's no like openwrt is using these partitions, i just want to let advanced users use them if they want.

> We can consider just using this extra space and split it to the 2 partition... (if the bootloader is ok with this)

as i see it, we have 4 options, 2 simple and 2 complex:

```
[    1.277588] Creating 15 MTD partitions on "qcom_nand.0":
[    1.284466] 0x000000000000-0x000000100000 : "SBL1"
[    1.291446] 0x000000100000-0x000000200000 : "MIBIB"
[    1.296083] 0x000000200000-0x000000300000 : "QSEE"
[    1.300798] 0x000000300000-0x000000380000 : "CDT"
[    1.305094] 0x000000380000-0x000000580000 : "APPSBL"
[    1.311471] 0x000000580000-0x000000600000 : "ART"
[    1.314994] 0x000000600000-0x000000680000 : "u_env"
[    1.319582] 0x000000680000-0x0000006c0000 : "s_env"
[    1.324016] 0x0000006c0000-0x000000700000 : "devinfo"
[    1.328823] 0x000000700000-0x00000a800000 : "kernel"
[    1.490871] 0x000000d00000-0x00000a800000 : "rootfs"
[    1.643151] mtd: setting mtd10 (rootfs) as root device
[    1.643537] mtdsplit: no squashfs found in "rootfs"
[    1.647219] 0x00000a800000-0x000014900000 : "alt_kernel"
[    1.809863] 0x00000ae00000-0x000014900000 : "alt_rootfs"
[    1.959552] 0x000014900000-0x000014b00000 : "sysdiag"
[    1.961963] 0x000014b00000-0x000020000000 : "syscfg"
```

1) make "sysdiag" and "syscfg" writable. we get one 181 MiB partition and a useless 2 MiB part. i put this option in draft mode for now

2) my recommended:
```
0x000014900000-0x000014b00000 : "stock_sysdiag"
0x000014b00000-0x000020000000 : "stock_syscfg"
0x000014900000-0x000020000000 : "extra"
```
we get one 183 MiB part, and access to return to stock if desired. i used this `stock_*` paradigm in https://openwrt.org/inbox/toh/arris/tr4400_v2, which required extensive partition rework to be able to coalesce most free space into a big UBI, including moving the irreplaceable `fw_env` paritition containing MACs, etc:
https://github.com/openwrt/openwrt/blob/bb8fd41f9a6775734882fe9707f7ad5e5edfbf9e/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts#L255-L258
https://github.com/openwrt/openwrt/blob/bb8fd41f9a6775734882fe9707f7ad5e5edfbf9e/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts#L272-L295

some space was not able to be coalesced because it had bootloader stuff in the way, so i defined an `extra` partition for it:
https://github.com/openwrt/openwrt/blob/bb8fd41f9a6775734882fe9707f7ad5e5edfbf9e/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts#L202-L205
https://github.com/openwrt/openwrt/blob/bb8fd41f9a6775734882fe9707f7ad5e5edfbf9e/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts#L300-L303

in all cases, although the part names were prefixed by `stock_`, the MTD numbers were conserved, so partitions could be accessed by scripts by number in both stock and openwrt firmwares.

again, i recommend this which sticks to these guidelines:
```
0x000014900000-0x000014b00000 : "stock_sysdiag"
0x000014b00000-0x000020000000 : "stock_syscfg"
0x000014900000-0x000020000000 : "extra"
```

3) adding the extra space to boot part set 2:

> We can consider just using this extra space and split it to the 2 partition... (if the bootloader is ok with this)

the firmware should be ok, as the kernel is raw outside of UBI, but the FW env and the openwrt build might need to be modified.

FW env:
```
altkern=a800000
auto_recovery=yes
baudrate=115200
boot_part_ready=3
bootcmd=if test $auto_recovery = no; then bootipq; elif test $boot_part = 1; then run bootpart1; else run bootpart2; fi
bootdelay=2
bootpart1=set bootargs $partbootargs && nand read $loadaddr $prikern $kernsize && bootm $loadaddr
bootpart2=set bootargs $partbootargs2 && nand read $loadaddr $altkern $kernsize && bootm $loadaddr
fdt_high=87000000
flash_type=2
flashimg=tftpboot $loadaddr $image && nand erase $prikern $imgsize && nand write $loadaddr $prikern $filesize
flashimg2=tftpboot $loadaddr $image && nand erase $altkern $imgsize && nand write $loadaddr $altkern $filesize
image=nodes_v2.img
imgsize=a100000
ipaddr=192.168.1.11
kernsize=d00000
loadaddr=81000000
machid=8010006
maxpartialboots=3
partbootargs=init=/sbin/init rootfstype=ubifs ubi.mtd=rootfs root=ubi0:ubifs rootwait rw
partbootargs2=init=/sbin/init rootfstype=ubifs ubi.mtd=alt_rootfs root=ubi0:ubifs rootwait rw 
prikern=700000
serverip=192.168.1.236
stderr=serial
stdin=serial
stdout=serial
boot_part=2
```

the problem here is that, during sysupgrade, kernel and UBI are written to the unused firmware slot and then `boot_part` is toggled. so what is the use of enlarging one UBI part while its alter ego stays the previous size? and disabling auto recovery means altering the fw variables, so this is a "complex" change.

4) if a complex change is desired (changing the build AND the FW env), then i would argue this is better:

change:

```
0x000000700000-0x00000a800000 : "kernel"
0x000000d00000-0x00000a800000 : "rootfs"
0x00000a800000-0x000014900000 : "alt_kernel"
0x00000ae00000-0x000014900000 : "alt_rootfs"
0x000014900000-0x000014b00000 : "sysdiag"
0x000014b00000-0x000020000000 : "syscfg"
```
to:
```
0x000000700000-0x00000a800000 : "stock_kernel"
0x000000d00000-0x00000a800000 : "stock_rootfs"
0x00000a800000-0x000014900000 : "stock_alt_kernel"
0x00000ae00000-0x000014900000 : "stock_alt_rootfs"
0x000014900000-0x000014b00000 : "stock_sysdiag"
0x000014b00000-0x000020000000 : "stock_syscfg"
0x000000700000-0x000001300000 : "recovery" - 12 MiB
0x000001300000-0x000001F00000 : "kernel" - 12 MiB
0x000001F00000-0x000020000000 : "ubi" - 481 MiB
```
or even to this if the BL supports kernel-in-ubi:
```
0x000000700000-0x00000a800000 : "stock_kernel"
0x000000d00000-0x00000a800000 : "stock_rootfs"
0x00000a800000-0x000014900000 : "stock_alt_kernel"
0x00000ae00000-0x000014900000 : "stock_alt_rootfs"
0x000014900000-0x000014b00000 : "stock_sysdiag"
0x000014b00000-0x000020000000 : "stock_syscfg"
0x000000700000-0x000001300000 : "recovery" - 12 MiB
0x000001300000-0x000020000000 : "ubi" - 493 MiB
```
recovery would be flashed with an initramfs build. the FW env would be patched to boot from ubi and switch to recovery on repeated boot failures (on boot_part set to 2). openwrt sysupgrade would be patched to always flash to and set boot_part 1 (as if always booted from boot_part 2).

recovery would also run as an immediate backup if ubi doesn't load, and there could be 3rd backup to recovery failing by trying to TFTP to ram in that case. [here](https://github.com/Lanchon/openwrt-tr4400-v2/blob/5fd99bd2712e6eb38b601f079ee0918d60643a16/install-recovery.sh#L78-L86) is an example of this (though there i also supported another recovery volume inside UBI, which could also be supported here).

to do this, i think a new build target would be needed, as well as an upgrade procedure to change the FW env. way too much complexity for a device with such little use. (plus i am not willing put the time to develop this myself; but still, this is how i think a complex change should be done.)

### so in the end...

i favor option 2) above: just a little change to the DTB that will go unnoticed to all except those wanting a little persistent storage on the device.

what do you think?

Lanchon commented:
btw @Ansuel, possible bug...

this scan should not be happening at all:
```
[    1.490871] 0x000000d00000-0x00000a800000 : "rootfs"
[    1.643151] mtd: setting mtd10 (rootfs) as root device
[    1.643537] mtdsplit: no squashfs found in "rootfs"
```

and on top it is happening in the wrong partition:
```
root@OpenWrt:~# fw_printenv -n boot_part
2
root@OpenWrt:~# cat /sys/class/ubi/ubi0/mtd_num 
12
root@OpenWrt:~# cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 00100000 00020000 "SBL1"
mtd1: 00100000 00020000 "MIBIB"
mtd2: 00100000 00020000 "QSEE"
mtd3: 00080000 00020000 "CDT"
mtd4: 00200000 00020000 "APPSBL"
mtd5: 00080000 00020000 "ART"
mtd6: 00080000 00020000 "u_env"
mtd7: 00040000 00020000 "s_env"
mtd8: 00040000 00020000 "devinfo"
mtd9: 0a100000 00020000 "kernel"
mtd10: 09b00000 00020000 "rootfs"
mtd11: 0a100000 00020000 "alt_kernel"
mtd12: 09b00000 00020000 "alt_rootfs"
mtd13: 00200000 00020000 "sysdiag"
mtd14: 0b500000 00020000 "syscfg"
```

and worse,  `rootfs` does contain a squashfs rootfs inside a volume. what if it finds it? the header is there for sure, what is stopping it from finding it?

Lanchon commented:
i updated the PR to option 2) above and took it out of draft mode/

Lanchon commented:
hi @dangowrt, any chance of merging this trivial change?
